### PR TITLE
Test cuda graph capture

### PR DESCRIPTION
**Pull request by gevtushenko**
## Description

closes https://github.com/NVIDIA/cccl/issues/321

This PR generalizes CDP wrapper to support multiple launch backends. 
It also introduces graph capture backend and adds coverage for graph capture to basic variants of device-scope tests. 

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/test/catch2_test_launch_helper.h [hunk 1]
```diff
@@ -40,11 +41,11 @@
 //!
 //! ```
 //! // Add PARAM to make CMake generate a test for both host and device launch:
-//! // %PARAM% TEST_CDP cdp 0:1
+//! // %PARAM% TEST_LAUNCH lid 0:1
```
Comment (miscco) [inline cub/test/catch2_test_launch_helper.h:44]: Should that go to 2 if we add the `#if TEST_LAUNCH == 2` below?

File: cub/test/catch2_test_launch_helper.h [hunk 2]
```diff
@@ -27,6 +27,7 @@
 
 #pragma once
 
+#include <stdexcept>
```
Comment (miscco) [inline cub/test/catch2_test_launch_helper.h:30]: Should we move that dependency down to the `#if TEST_LAUNCH == 2` block?

File: cub/test/catch2_test_launch_wrapper.cu [hunk 3]
```diff
@@ -113,34 +121,34 @@ struct cub_api_example_t
...
   }
 };
 
-DECLARE_CDP_WRAPPER(cub_api_example_t::x2_0, x2_0);
-DECLARE_CDP_WRAPPER(cub_api_example_t::x0_5, x0_5);
+DECLARE_LAUNCH_WRAPPER(cub_api_example_t::x2_0, x2_0);
+DECLARE_LAUNCH_WRAPPER(cub_api_example_t::x0_5, x0_5);
 
-CUB_TEST("CDP wrapper works with predefined invocables", "[test][utils]")
+CUB_TEST("Laucn wrapper works with predefined invocables", "[test][utils]")
```
Comment (miscco) [inline cub/test/catch2_test_launch_wrapper.cu:139]:
```suggestion
CUB_TEST("Launch wrapper works with predefined invocables", "[test][utils]")
```

File: cub/test/catch2_test_launch_wrapper.cu [hunk 4]
```diff
@@ -67,8 +67,10 @@ struct cub_api_example_t
...
                                                  T *d_out,
                                                  int num_items,
-                                                 bool should_be_invoked_on_device)
+                                                 cudaStream_t stream = 0)
   {
+    constexpr bool should_be_invoked_on_device = TEST_LAUNCH == 1;
+
     NV_IF_TARGET(NV_IS_HOST,
                  (if (should_be_invoked_on_device) { return cudaErrorLaunchFailure; }),
                  (if (!should_be_invoked_on_device) { return cudaErrorLaunchFailure; }));
```
Comment (miscco) [inline cub/test/catch2_test_launch_wrapper.cu:76]: That line of code is super confusing as it is difficult to discern that this is actually an NV_IF_ELSE_TARGET

Could you change the `NV_IF_TARGET` to `NV_IF_ELSE_TARGET`

File: cub/test/catch2_test_launch_helper.h [hunk 5]
```diff
@@ -59,11 +60,70 @@
...
 
-#if !defined(TEST_CDP)
-#error Test file should contain %PARAM% TEST_CDP cdp 0:1
+#if !defined(TEST_LAUNCH)
+#error Test file should contain %PARAM% TEST_LAUNCH lid 0:1
 #endif
 
-#if TEST_CDP == 1
+
+#if TEST_LAUNCH == 2
```
Comment (miscco) [inline cub/test/catch2_test_launch_helper.h:68]: proposal no change requested:

At some point in the near future we should move this from numerical constants to enumerations with a proper name

**Review by miscco — APPROVED** [review]
Looks great! 

Only some minor nits

**Review by allisonvacanti — APPROVED** [review]
Really cool stuff, impressively small patch for the amount of coverage this adds!

Some minor cleanups / discussion points inline, but this lgtm.

File: cub/test/CMakeLists.txt [hunk 6]
```diff
@@ -129,6 +129,17 @@ function(cub_is_catch2_test result_var test_src)
   endif()
 endfunction()
 
+## cub_cdp_by_lid
+#
+# If given launcher id corresponds to a CDP launcher, set `out_var` to 1.
+function(cub_cdp_by_lid out_var launcher_id)
```
Comment (allisonvacanti) [inline cub/test/CMakeLists.txt:135]: Suggestion: rename this function to `_cub_launcher_requires_rdc`.

File: cub/test/CMakeLists.txt [hunk 7]
```diff
@@ -348,31 +362,33 @@ foreach (test_src IN LISTS test_srcs)
...
 
         # Enable RDC if the test either:
-        # 1. Explicitly requests it (cdp_1 label)
-        # 2. Does not have an explicit CDP variant (no cdp_0 or cdp_1) but
+        # 1. Explicitly requests it (lid_1 label)
+        # 2. Does not have an explicit CDP variant (no lid_0, lid_1, or lid_2) but
         #    RDC testing is forced
         #
-        # Tests that explicitly request no cdp (cdp_0 label) should never enable
+        # Tests that explicitly request no cdp (lid_0 label) should never enable
```
Comment (allisonvacanti) [inline cub/test/CMakeLists.txt:389]: What about graph capture + RDC? Is device-side graph capture supported, and if so, should we have coverage?

File: cub/test/catch2_test_device_adjacent_difference_substract_left.cu [hunk 8]
```diff
@@ -35,13 +35,13 @@
...
+#include "catch2_test_launch_helper.h"
 #include "catch2_test_helper.h"
 
-DECLARE_CDP_WRAPPER(cub::DeviceAdjacentDifference::SubtractLeft, adjacent_difference_subtract_left);
-DECLARE_CDP_WRAPPER(cub::DeviceAdjacentDifference::SubtractLeftCopy, adjacent_difference_subtract_left_copy);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceAdjacentDifference::SubtractLeft, adjacent_difference_subtract_left);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceAdjacentDifference::SubtractLeftCopy, adjacent_difference_subtract_left_copy);
 
-// %PARAM% TEST_CDP cdp 0:1
+// %PARAM% TEST_LAUNCH lid 0:1:2
```
[the remark this answers is not in the document]
Comment (gevtushenko) [reply cub/test/catch2_test_device_adjacent_difference_substract_left.cu:44]: CDPv2 is tail kernel launch. We do not provide kernel-level API, so there'll be no CDPv2 tests.

File: cub/test/CMakeLists.txt (hunk 6 above)
Comment (allisonvacanti) [reply cub/test/CMakeLists.txt:135]: The intention of the underscore was to distinguish between "heavy" functions that might be used repeatedly across multiple files as a sort of API vs. helper functions that just simplify code flow in our implementations (like these one-off functions that don't do anything interesting except map one value into another to make life a little easier).

This was more important for thrust, which exposes some functions in the cmake package (`thrust_create_target`) that are actual user APIs. My goal was to have the more useful functions appear in IDE autocompletion while the helpers wouldn't clutter up the list of completions. It doesn't have any actual impact on correctness, though, it was just a convention.

In this suggestion, the new name was more important with the underscore. `launcher_requires_rdc` makes it clear that this is used solely to map launcher values to RDC flags.

File: cub/test/catch2_test_launch_helper.h (hunk 1 above)
Comment (gevtushenko) [reply cub/test/catch2_test_launch_helper.h:44]: I was hesitant to add it here, not to make graph capture covered by default. But I think it's a good idea to at least document it. I'll add a section below.

File: cub/test/CMakeLists.txt (hunk 6 above)
Comment (gevtushenko) [reply cub/test/CMakeLists.txt]: Thank you for elaborating on this! I've changed the function name per recommendation. The convention doesn't seem intuitive. If I'm not missing anything, it's not documented. Maybe it's worth adding cmake developer overview, describing it.

**Pull request merged**